### PR TITLE
EG-554, FactCheck accept/reject Controller/Service/View/Connector

### PR DESCRIPTION
**Pull request by cmrhmrc**

### CODE REVIEW

File: app/controllers/FactCheckPageReviewController.scala [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  def onSubmit(processId: String, page: String): Action[AnyContent] = factCheckerReviewerIdentifierAction.async { implicit request =>
+    val form: Form[FactCheckPageReview] = formProvider()
+    form
+      .bindFromRequest()
+      .fold(
+        (formWithErrors: Form[FactCheckPageReview]) => { Future.successful(BadRequest(view(processId, s"/$page", formWithErrors))) },
+        result => {
+          val reviewDetail = PageReviewDetail(processId, s"/$page", Some(result.answer), Complete, updateUser = Some(s"${request.credId}:${request.name}"))
```
Comment (woollett-r) [inline app/controllers/FactCheckPageReviewController.scala:72]: I'm not certain we need the option on the user details. With the way the identifier action works there should always be value for the user name and credentials. If the values are not available the identifier should raise an error.

File: app/controllers/FactCheckPageReviewController.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  def onPageLoad(processId: String, page: String): Action[AnyContent] = factCheckerReviewerIdentifierAction.async { implicit request =>
+    reviewService.factCheckPageInfo(processId, s"/$page") map {
+      case Right(data) if data.result.isDefined =>
+        val form: Form[FactCheckPageReview] = formProvider().bind(Map("answer" -> data.result.fold("")(_.toString)))
+        Ok(view(processId, s"/$page", form))
+
+      case Right(_) => Ok(view(processId, s"/$page", formProvider()))
+      case Left(err) =>
+        // Handle stale data, internal server and any unexpected errors
+        logger.error(s"Request for approval 2i page review for process $processId and page $page returned error $err")
```
Comment (woollett-r) [inline app/controllers/FactCheckPageReviewController.scala:60]: Fact check not 2i.

File: app/controllers/FactCheckPageReviewController.scala [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    form
+      .bindFromRequest()
+      .fold(
+        (formWithErrors: Form[FactCheckPageReview]) => { Future.successful(BadRequest(view(processId, s"/$page", formWithErrors))) },
+        result => {
+          val reviewDetail = PageReviewDetail(processId, s"/$page", Some(result.answer), Complete, updateUser = Some(s"${request.credId}:${request.name}"))
+          reviewService.factCheckPageComplete(processId, s"/$page", reviewDetail).map {
+            case Right(_) => Redirect(routes.FactCheckController.approval(processId))
+            case Left(NotFoundError) =>
+              logger.error(s"Unable to retrieve approval 2i page review for process $processId")
```
Comment (woollett-r) [inline app/controllers/FactCheckPageReviewController.scala:76]: Fact check.

File: app/controllers/FactCheckPageReviewController.scala [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+              NotFound(errorHandler.notFoundTemplate)
+            case Left(StaleDataError) =>
+              logger.warn(s"The requested approval 2i review for process $processId can no longer be found")
```
Comment (woollett-r) [inline app/controllers/FactCheckPageReviewController.scala:79]: Fact check

File: app/controllers/FactCheckPageReviewController.scala [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+              NotFound(errorHandler.notFoundTemplate)
+            case Left(err) =>
+              // Handle internal server and any unexpected errors
+              logger.error(s"Request for approval 2i review process for process $processId returned error $err")
```
Comment (woollett-r) [inline app/controllers/FactCheckPageReviewController.scala:83]: Fact check

File: app/controllers/FactCheckPageReviewController.scala (hunk 1 above)
Comment (cmrhmrc) [reply app/controllers/FactCheckPageReviewController.scala:72]: Yes, this is a bit of history which goes all the way through both 2i and factcheck and down to external-guidance. Could create bug ticket to resolve across all microservices if deemed necessary

File: app/controllers/FactCheckPageReviewController.scala (hunk 2 above)
Comment (cmrhmrc) [reply app/controllers/FactCheckPageReviewController.scala:60]: will update

File: app/controllers/FactCheckPageReviewController.scala (hunk 3 above)
Comment (cmrhmrc) [reply app/controllers/FactCheckPageReviewController.scala:76]: will update

File: app/controllers/FactCheckPageReviewController.scala (hunk 4 above)
Comment (cmrhmrc) [reply app/controllers/FactCheckPageReviewController.scala:79]: will update

File: app/controllers/FactCheckPageReviewController.scala (hunk 5 above)
Comment (cmrhmrc) [reply app/controllers/FactCheckPageReviewController.scala:83]: will update

File: app/views/fact_check_content_review.scala.html [hunk 6]
```diff
@@ -48,7 +48,7 @@ <h2 class="app-task-list__section">
         @for((page, index) <- review.pages.zipWithIndex) {
           <li class="app-task-list__item">
             <span class="app-task-list__task-name">
-              <a href="#" aria-describedby="page-@index">@page.title</a>
+              <a href="@{routes.FactCheckPageReviewController.onPageLoad(review.ocelotId, page.title.drop(1)).toString}" aria-describedby="page-@index">@page.title</a>
```
Comment (woollett-r) [inline app/views/fact_check_content_review.scala.html:51]: You could also use substring(1), but that would be resorting to Java!

Comment (cmrhmrc) [reply app/views/fact_check_content_review.scala.html:51]: .... did someone mention Java ????

File: conf/messages.en [hunk 7]
```diff
@@ -58,6 +58,16 @@ factCheck.pagesReviewedConfirmationHeading = Confirm you''ve fact checked all pa
 factCheck.sendConfirmation = Confirm
 factCheck.sendConfirmationStatus =
 
+factCheckPageReview.answer.error.required = You must select one of the options shown
+factCheckPageReview.comment.error.required = No comment supplied
```
Comment (woollett-r) [inline conf/messages.en:62]: We don't need to worry about this now, but I think the comments may be optional.

File: test/controllers/FactCheckPageReviewControllerSpec.scala [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+      MockReviewService
+        .factCheckPageComplete(id, updatedReviewDetail.pageUrl, updatedReviewDetail)
+        .returns(Future.successful(Right(())))
+
+      val result: Future[Result] = reviewController.onSubmit(id, updatedReviewDetail.pageUrl.drop(1))(fakeRequest.withFormUrlEncodedBody(("answer", "Yes")))
+
+      status(result) shouldBe Status.SEE_OTHER
+    }
+
+    "Return an Html document displaying the details of the review result" in new Test {
```
Comment (woollett-r) [inline test/controllers/FactCheckPageReviewControllerSpec.scala:99]: Can it really be an HTML document if the content type is None?

File: test/forms/FactCheckPageReviewFormProviderSpec.scala [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  private trait Test {
+
+    val formProvider = new FactCheckPageReviewFormProvider()
+
+    val form: Form[FactCheckPageReview] = formProvider()
+
+    val answer = "answer"
+    val comment = "comment"
+
+    val definedComment = "This page is acceptable"
```
Comment (woollett-r) [inline test/forms/FactCheckPageReviewFormProviderSpec.scala:35]: We should really remove the comment related bits here. It is interesting that the tests still work!

File: test/forms/FactCheckPageReviewFormProviderSpec.scala [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    val emptyComment = ""
+
+    val requiredAnswerKeyError = "factCheckPageReview.answer.error.required"
+    val requiredAnswerKeyFormError = FormError(answer, requiredAnswerKeyError)
+
+    val requiredCommentKeyError = "factCheckPPageReview.comment.error.required"
+    val requiredCommentKeyFormError = FormError(comment, requiredCommentKeyError)
+  }
+
+  "The two eye page review form provider" should {
```
Comment (woollett-r) [inline test/forms/FactCheckPageReviewFormProviderSpec.scala:45]: Fact check

File: conf/messages.en (hunk 7 above)
Comment (cmrhmrc) [reply conf/messages.en:62]: Intentionally left comment related items which dont affect the running of the app

File: test/controllers/FactCheckPageReviewControllerSpec.scala (hunk 8 above)
Comment (cmrhmrc) [reply test/controllers/FactCheckPageReviewControllerSpec.scala:99]: Odd, this a copy of the 2i page review tests and that passes also

File: test/forms/FactCheckPageReviewFormProviderSpec.scala (hunk 9 above)
Comment (cmrhmrc) [reply test/forms/FactCheckPageReviewFormProviderSpec.scala:35]: the non-impact comment related stuff intentionally left in. The form bind ignores what it doesnt want

File: test/forms/FactCheckPageReviewFormProviderSpec.scala (hunk 10 above)
Comment (cmrhmrc) [reply test/forms/FactCheckPageReviewFormProviderSpec.scala:45]: will update

File: test/controllers/FactCheckPageReviewControllerSpec.scala (hunk 8 above)
Comment (woollett-r) [reply test/controllers/FactCheckPageReviewControllerSpec.scala:99]: As the response is a redirect there is probably only a status code and location. As such there is no Html body. Should really change the description.

**Pull request merged**